Make the code a bit more pythonic

jvoisin opened the pull request:
Do feel free to ship me a Binja license so I can test this PR ;)

jvoisin committed:
Make the code a bit more pythonic

Do feel free to ship me a Binja license so I can test this PR ;)

paulfariello commented on binja-plugins/djumpo_unchained/__init__.py:
```diff
@@ -37,15 +37,9 @@ def flatten(self, address):
-				self.todo(edges[1].target.start)
+
+                        for edge in block.outgoing_edges:
+                                self._todo.append(edge.target.start)
```
You removed an important comment. Here we have to ensure we follow the immediate branch first. Otherwise this is clearly more pythonic and more readable :+1:

paulfariello commented on binja-plugins/djumpo_unchained/__init__.py:
```diff
@@ -54,7 +48,7 @@ def process_block(self, block):
 
 		# Add label
-		if len(block.incoming_edges):
+		if block.incoming_edges:
```
I'm not found of removing the `len()`. Now when reading the code we are not sure what we really expect. An empty list? `False`? An empty string? `0`?

jvoisin replied:
This is the [Python way]( https://www.python.org/dev/peps/pep-0008/ )™©®. The rationale behind this is that `__bool__` is the interface for testing emptiness: all default types are implementing it (you can get its return by casting your object with `bool(your_object)`. Inside a condition, Python is implicitely casting to bool. You can have an object that is empty but `True`.

But yeah, I do agree that it's a bit annoying. Feel free to skip this part of the PR ;)

paulfariello commented on binja-plugins/djumpo_unchained/__init__.py:
```diff
@@ -89,12 +80,12 @@ def process_inst(self, inst, block):
 		for token in tokens:
 			symbol = self._view.get_symbol_by_raw_name(token.text)
-			if symbol is not None:
+			if symbol:
```
Same comment as for the `if len(`

paulfariello commented on binja-plugins/djumpo_unchained/__init__.py:
```diff
@@ -26,7 +26,7 @@ def todo(self, address):
 	def flatten(self, address):
 		code = ''
-		self.todo(address)
+		self._todo.append(address)
```
Why would you skip the todo function?

jvoisin replied:
because there is no point in having a function to simply do an `append` to a list in my opinion.

paulfariello commented on binja-plugins/djumpo_unchained/__init__.py:
```diff
@@ -89,12 +80,12 @@ def process_inst(self, inst, block):
 		edges = block.outgoing_edges
-		nb_edges = len(edges)
-		if nb_edges > 0 and inst_name in JUMPS:
+		if edges > 0 and inst_name in JUMPS:
```
`[] > 0`, not sure that's valid python 3. And since python 2 will retire in 1 Year 9 Months 11 Days 10 Hours we should better keep the `len()`.

jvoisin replied:
Err, this is a typo

paulfariello reviewed (commented):
Glad to have you as our first contributor :wink: 
Maybe we should ask @Vector35 to get some free licences.

Meanwhile here are a few comments.

jvoisin committed:
Fix a typo

jvoisin commented:
Do feel free to throw one license at me if you do so ;)

You should also use `set` instead of `list`, if possible, to reduce runtime complexity. I didn't include this in my PR, since I have no idea if offsets are hashable, although I would be surprised if they weren't.

paulfariello commented:
Argh just realize we commit this script with tabs… So your PR is mixing space and tabs.
We will fix indentation on master later.

jvoisin commented:
Trow pep8 at it!

paulfariello committed:
Remove unused function

paulfariello committed:
Add useful comment

Order of flattened block is really important since immediate branch must
directly follow previous block

paulfariello committed:
We prefer explicit if statement

paulfariello commented:
Successfully tested with our BinaryNinja licence!

Thanks @jvoisin

jvoisin commented:
Care to try to use `set` instead of `liste` [here]( https://github.com/ThisIsSecurity/sinkhole/blob/master/binja-plugins/djumpo_unchained/__init__.py#L21 )?

paulfariello commented:
Yeah but I'll do it in another branch.

jvoisin commented:
![]( https://media.giphy.com/media/l2Sq5GffrCyUMEXjW/giphy.gif )

paulfariello commented:
Done in 895cb21460ffb93f1cc382e6ac3b16db85ccfdf5
Address are simple integers, so they are hashable.

jvoisin commented:
You can do the same with `_todo` ;)

paulfariello commented:
`_todo` must be a lifo (the immediate branch stuff) so I can't.

jvoisin commented:
ho, right :/
